nntrainer/tensor/cl_operations/cl_kernels/int4_gemv.cl

```diff
@@ -160,6 +235,59 @@ DECLARE_BLOCK_READ_EMULATION(1, 16)
   BLOCK_READN(half, INPUT_TILE_SIZE, ptr, offset)
 #define GEMV_FILTER_BLOCK_READ(ptr, offset) BLOCK_READN(char, 16, ptr, offset)
 
+// ==========================================================================
+// Non-Intel subgroup broadcast / reduce emulation
+//   - Subgroups are lanes along X (local_id(0))
+//   - thr_id is Z-dimension (local_id(2))
+//   - Each thr_id slice gets its own 16-element buffer
+// ==========================================================================
+#if !defined(cl_intel_subgroups)
+
+inline float sg_reduce_add_float(float v, __local float *buf_line) {
+  uint lid = get_sub_group_local_id(); // lane in X-dimension
+  buf_line[lid] = v;
+  barrier(CLK_LOCAL_MEM_FENCE);
+
+  uint sg_size = SUBGROUP_SIZE; // expected 16
+  for (uint stride = sg_size >> 1; stride > 0; stride >>= 1) {
```

Comment:
I don't get it:

shouldn't
```
if (get_subgroup_local_id() != 0)
{
   for (uint i = 1; i <SUBGROUP_SIZE; ++i)
   {
       buf_line[0] = buf_line[0] + buf_line[i];
   }
}
```
as we would want to reduce all no just some.
Simply don't get it what's going on with this stride.

Reply:
Oh think I get it, to make it work the way you want the reduction accumulation local buf_line[0] + to strided element should be atomic (local atomit fetch add) otherwise it races between subgroup threads. (race between is on buf_line[0] and modifiaction of buf_line[0])

Have you tested correctness of this reduction?